design-proposals/vm-delete-protection.md

```diff
@@ -0,0 +1,171 @@
+# Overview
+
+VirtualMachine Delete Protection prevents the accidental deletion of VMs from a cluster. As long as deletion protection
+is activated for certain VM objects, they cannot be deleted unless the protection is intentionally removed.
+
+## Motivation
+
+KubeVirt clusters are often managed by automations, CLI commands, 3rd party tools or UI actions, for example to deploy,
+reconfigure or delete VMs. These automations may result in accidental deletion of VMs that should have not been deleted.
+Affected VMs could be business critical to application/s or services the cluster is running, this could cause them a
+degradation or out of service, e.g., a SQL server is not available to handle requests.
+Moreover, the affected VMs may contain crucial data if underlining PVC is deleted as a result of a cascaded delete. 
+Therefore, the VM deletion protection aims to avoid accidental degradation of applications or loss of data.
+
+## Goals
+
+- Create a delete protection for VirtualMachine objects.
+
+## Non Goals
+
+- Create a delete protection for VirtualMachineInstance objects.
+- Create a delete protection for virt-launcher Pod objects.
+
+## Definition of Users
+
+VM owners with permissions to delete VMs.
+Cluster administrators with permissions to delete VMs.
+
+## User Stories
+
+As a user who has the permissions to delete VMs, I want to protect my VMs from accidental deletion.
+
+## Repos
+
+kubevirt/kubevirt
+
+# Design
+
+The design is divided into two sections. The first section shows how the VM protection can be enabled by a user, the
+second section depicts possible implementations of the feature in the backend.
+
+### Enable the VM Object Protection
+
+To protect VMs against deletion, they need to be annotated with  `kubevirt.io/vm-delete-protection` set to `true` or
+`True`. If the annotation is present with one of the aforementioned values, requests to delete this VM will be rejected
+by KubeVirt. By default, this annotation is not present on a VM. It is up to the user to enable the deletion protection.
+The annotation looks like this in the VM manifests:
+
+```yaml
+apiVersion: kubevirt.io/v1
+kind: VirtualMachine
+metadata:
+  annotations:
+    kubevirt.io/vm-delete-protection: "true"
+spec:
+[...]
+```
+
+### Disable the VM Object Protection
+There are two possible ways to disable the VM protection:
+
+1. Removing the `kubevirt.io/vm-delete-protection` annotation with the following patch operation:
+```bash
+$ kubectl patch vm <vm-name> --type=json -p='[{"op": "remove", "path": "/metadata/annotations/kubevirt.io~1vm-delete-protection"}]'
```

Comment:
This ultimately means that anyone with the ability to update/patch the VM can enable or disable this functionality. Have we considered placing it in status controlled by a subresource with finer grain RBAC contol?

Reply:
Correct me if I'm wrong, but I assume this is the same concern raised in the [top comment of your review](https://github.com/kubevirt/community/pull/363#pullrequestreview-2491965394).

I didn't think about that TBH, but it feels a bit overkill IMHO.  However, what would be the ultimate goal of doing that? I'm not sure if this particular use case requires it. Please, elaborate a bit if you think otherwise.

In any case, if we find out this is required, we could implement it in a follow-up.

Reply:
The ultimate goal would be to allow a smaller subset of users (super admins?) to have access to the delete protection toggle than have the ability to update the VM object itself. It's extra work but given the use case here I'm not sure we want to give everyone access to this toggle right?

Reply:
@dominikholler well it's any user with DELETE privileges against the VM surely so not just an owner but anyone in the same namespace or admin. It also doesn't protect against that same user accidentently removing the label and thus the protection before requesting the deletion.

Moving the toggle into the status would avoid this as users wouldn't be able to touch it without going through the sub resource API that itself could be limited to a specific owner and/or super set of admins.

Reply:
>  It also doesn't protect against that same user accidentently removing the label and thus the protection before requesting the deletion.

This would require two steps, this feature is just protecting from a single unintended step, e.g. a `kubectl delete vm/xxx" .